File: leetcode022.py
class Solution:
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        if n == 1:
            return ['()']
        res = []
        t = self.generateParenthesis(n-1)
        for x in t:
            h = '(' + x
            for i in range(len(h)):
                if h[i] == '(':
                    res.append(h[:i+1]+')'+h[i+1:])
        res = set(res)
        res = list(res)
        return res

File: leetcodec++/leetcode022.cpp
//
// Created by 黄敏辉 on 2019-06-28.
//

class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        helper(res, "", n, 0);
        return res;
    }

    void helper(vector<srting> &res, string s, l, r){
        if (l == 0 && r == 0){
            res.push_back(s);
            return;
        }
        if (l > 0)
            helper(res, s + "(", l - 1, r + 1);
        if (r > 0)
            helper(res, s + ")", l , r - 1);
    }
};